File: client/src/test/scala/stasis/test/specs/unit/client/tracking/trackers/DefaultTrackerSpec.scala
package stasis.test.specs.unit.client.tracking.trackers

import akka.actor.typed.scaladsl.Behaviors
import akka.actor.typed.{ActorSystem, Behavior, SpawnProtocol}
import akka.stream.scaladsl.Sink
import org.scalatest.concurrent.Eventually
import org.scalatest.{Assertion, BeforeAndAfterAll}
import stasis.client.collection.rules.Rule
import stasis.client.tracking.TrackerView
import stasis.client.tracking.trackers.DefaultTracker
import stasis.core.persistence.backends.memory.EventLogMemoryBackend
import stasis.shared.model.datasets.DatasetEntry
import stasis.shared.ops.Operation
import stasis.test.specs.unit.AsyncUnitSpec
import stasis.test.specs.unit.core.telemetry.MockTelemetryContext

import java.nio.file.Paths
import scala.collection.immutable.Queue
import scala.concurrent.duration._

class DefaultTrackerSpec extends AsyncUnitSpec with Eventually with BeforeAndAfterAll {
  "A DefaultTracker" should "track backup events" in {
    val tracker = createTracker()

    implicit val operation: Operation.Id = Operation.generateId()
    val file = Paths.get("test").toAbsolutePath

    val initialState = tracker.state.await

    initialState should be(
      TrackerView.State.empty
    )

    val rule = Rule(
      operation = Rule.Operation.Include,
      directory = "/work",
      pattern = "?",
      comment = None,
      original = Rule.Original(line = "", lineNumber = 0)
    )

    tracker.backup.entityDiscovered(entity = file)
    tracker.backup.entityDiscovered(entity = file)
    tracker.backup.entityDiscovered(entity = file)
    tracker.backup.entityDiscovered(entity = file)
    tracker.backup.specificationProcessed(unmatched = Seq.empty)
    tracker.backup.specificationProcessed(unmatched = Seq(rule -> new RuntimeException("Test failure")))
    tracker.backup.entityExamined(entity = file, metadataChanged = true, contentChanged = false)
    tracker.backup.entityExamined(entity = file, metadataChanged = true, contentChanged = true)
    tracker.backup.entityCollected(entity = file)
    tracker.backup.entityProcessed(entity = file, contentChanged = true)
    tracker.backup.metadataCollected()
    tracker.backup.metadataPushed(entry = DatasetEntry.generateId())
    tracker.backup.failureEncountered(failure = new RuntimeException("test failure"))
    tracker.backup.completed()

    eventually[Assertion] {
      val completedState = tracker.state.await.operations(operation)

      completedState.completed should not be empty

      completedState.stages.contains("discovery") should be(true)
      completedState.stages("discovery").steps.size should be(4)

      completedState.stages.contains("specification") should be(true)
      completedState.stages("specification").steps.size should be(1)

      completedState.stages.contains("examination") should be(true)
      completedState.stages("examination").steps.size should be(2)

      completedState.stages.contains("collection") should be(true)
      completedState.stages("collection").steps.size should be(1)

      completedState.stages.contains("processing") should be(true)
      completedState.stages("processing").steps.size should be(1)

      completedState.stages.contains("metadata") should be(true)
      completedState.stages("metadata").steps.size should be(2)

      completedState.failures should be(
        Queue(
          "RuleMatchingFailure: Rule [+ /work ?] failed with [RuntimeException - Test failure]",
          "RuntimeException: test failure"
        )
      )
    }
  }

  it should "track recovery events" in {
    val tracker = createTracker()

    implicit val operation: Operation.Id = Operation.generateId()
    val file = Paths.get("test").toAbsolutePath

    val initialState = tracker.state.await

    initialState should be(
      TrackerView.State.empty
    )

    tracker.recovery.entityExamined(entity = file, metadataChanged = true, contentChanged = false)
    tracker.recovery.entityExamined(entity = file, metadataChanged = true, contentChanged = true)
    tracker.recovery.entityCollected(entity = file)
    tracker.recovery.entityProcessed(entity = file)
    tracker.recovery.metadataApplied(entity = file)
    tracker.recovery.failureEncountered(failure = new RuntimeException("test failure"))
    tracker.recovery.completed()

    eventually[Assertion] {
      val completedState = tracker.state.await.operations(operation)

      completedState.completed should not be empty

      completedState.stages.contains("examination") should be(true)
      completedState.stages("examination").steps.size should be(2)

      completedState.stages.contains("collection") should be(true)
      completedState.stages("collection").steps.size should be(1)

      completedState.stages.contains("processing") should be(true)
      completedState.stages("processing").steps.size should be(1)

      completedState.stages.contains("metadata-applied") should be(true)
      completedState.stages("metadata-applied").steps.size should be(1)

      completedState.failures should be(Queue("RuntimeException: test failure"))
    }
  }

  it should "track server events" in {
    val tracker = createTracker()

    val server1 = "test-server-01"
    val server2 = "test-server-02"

    val initialState = tracker.state.await

    initialState should be(
      TrackerView.State.empty
    )

    tracker.server.reachable(server1)
    tracker.server.reachable(server2)
    tracker.server.unreachable(server1)

    eventually[Assertion] {
      tracker.state.await.servers.view.mapValues(_.reachable).toMap should be(
        Map(
          server1 -> false,
          server2 -> true
        )
      )
    }
  }

  it should "provide state updates" in {
    val tracker = createTracker()

    implicit val operation: Operation.Id = Operation.generateId()
    val file = Paths.get("test").toAbsolutePath

    val initialState = tracker.state.await

    initialState should be(
      TrackerView.State.empty
    )

    val expectedUpdates = 3
    val updates = tracker.stateUpdates.take(expectedUpdates.toLong).runWith(Sink.seq)

    tracker.backup.entityExamined(entity = file, metadataChanged = true, contentChanged = false)
    await(50.millis, withSystem = system)
    tracker.backup.entityCollected(entity = file)
    await(50.millis, withSystem = system)
    tracker.recovery.completed()

    updates.await.flatMap(_.operations.get(operation)).toList match {
      case first :: second :: third :: Nil =>
        first.stages.size should be(1)
        first.completed should be(empty)

        second.stages.size should be(2)
        second.completed should be(empty)

        third.stages.size should be(2)
        third.completed should not be empty

      case other =>
        fail(s"Unexpected result received: [$other]")
    }
  }

  it should "provide operation updates" in {
    val tracker = createTracker()

    implicit val operation: Operation.Id = Operation.generateId()
    val file = Paths.get("test").toAbsolutePath

    val initialState = tracker.state.await

    initialState should be(
      TrackerView.State.empty
    )

    val expectedUpdates = 2
    val updates = tracker.operationUpdates(operation).take(expectedUpdates.toLong).runWith(Sink.seq)

    tracker.backup.entityExamined(entity = file, metadataChanged = true, contentChanged = false)
    await(50.millis, withSystem = system)
    tracker.backup.entityCollected(entity = file)(operation = Operation.generateId()) // other operation
    await(50.millis, withSystem = system)
    tracker.recovery.completed()

    updates.await.toList match {
      case first :: second :: Nil =>
        first.stages.size should be(1)
        first.completed should be(empty)

        second.stages.size should be(1)
        second.completed should not be empty

      case other =>
        fail(s"Unexpected result received: [$other]")
    }
  }

  private def createTracker(): DefaultTracker =
    DefaultTracker(
      createBackend = state =>
        EventLogMemoryBackend(
          name = s"test-tracker-${java.util.UUID.randomUUID()}",
          initialState = state
        )
    )

  override implicit val patienceConfig: PatienceConfig = PatienceConfig(5.seconds, 250.milliseconds)

  private implicit val system: ActorSystem[SpawnProtocol.Command] = ActorSystem(
    Behaviors.setup(_ => SpawnProtocol()): Behavior[SpawnProtocol.Command],
    "DefaultTrackerSpec"
  )

  private implicit val telemetry: MockTelemetryContext = MockTelemetryContext()

  override protected def afterAll(): Unit =
    system.terminate()
}

File: client-android/app/src/androidTest/java/stasis/client_android/tracking/DefaultTrackerSpec.kt
package stasis.client_android.tracking

import androidx.arch.core.executor.testing.InstantTaskExecutorRule
import androidx.test.ext.junit.runners.AndroidJUnit4
import kotlinx.coroutines.runBlocking
import org.hamcrest.CoreMatchers.equalTo
import org.hamcrest.CoreMatchers.not
import org.hamcrest.MatcherAssert.assertThat
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith
import stasis.client_android.await
import stasis.client_android.eventually
import stasis.client_android.lib.ops.Operation
import stasis.client_android.lib.ops.OperationId
import stasis.client_android.notNull
import java.nio.file.Paths
import java.util.UUID

@RunWith(AndroidJUnit4::class)
class DefaultTrackerSpec {
    @Test
    fun trackBackupEvents() {
        val tracker = DefaultTracker()

        val operation: OperationId = UUID.randomUUID()
        val file = Paths.get("test").toAbsolutePath()

        val initialState = tracker.state.value

        assertThat(initialState, equalTo(TrackerView.State.empty()))

        tracker.backup.entityDiscovered(operation, entity = file)
        tracker.backup.entityDiscovered(operation, entity = file)
        tracker.backup.entityDiscovered(operation, entity = file)
        tracker.backup.entityDiscovered(operation, entity = file)
        tracker.backup.specificationProcessed(operation, unmatched = emptyList())
        tracker.backup.entityExamined(operation, entity = file, metadataChanged = true, contentChanged = false)
        tracker.backup.entityExamined(operation, entity = file, metadataChanged = true, contentChanged = true)
        tracker.backup.entityCollected(operation, entity = file)
        tracker.backup.entityProcessed(operation, entity = file, contentChanged = true)
        tracker.backup.metadataCollected(operation)
        tracker.backup.metadataPushed(operation, entry = UUID.randomUUID())
        tracker.backup.failureEncountered(operation, failure = RuntimeException("test failure"))
        tracker.backup.completed(operation)

        runBlocking {
            eventually {
                val completedState = tracker.state.await().operations[operation].notNull()

                assertThat(completedState.completed != null, equalTo(true))

                assertThat(completedState.stages.contains("discovery"), equalTo(true))
                assertThat(completedState.stages["discovery"]!!.steps.size, equalTo(4))

                assertThat(completedState.stages.contains("specification"), equalTo(true))
                assertThat(completedState.stages["specification"]!!.steps.size, equalTo(1))

                assertThat(completedState.stages.contains("examination"), equalTo(true))
                assertThat(completedState.stages["examination"]!!.steps.size, equalTo(2))

                assertThat(completedState.stages.contains("collection"), equalTo(true))
                assertThat(completedState.stages["collection"]!!.steps.size, equalTo(1))

                assertThat(completedState.stages.contains("processing"), equalTo(true))
                assertThat(completedState.stages["processing"]!!.steps.size, equalTo(1))

                assertThat(completedState.stages.contains("metadata"), equalTo(true))
                assertThat(completedState.stages["metadata"]!!.steps.size, equalTo(2))

                assertThat(completedState.failures, equalTo(listOf("test failure")))
            }
        }
    }

    @Test
    fun trackBackupEventsWithUnmatchedRules() {
        val tracker = DefaultTracker()

        val operation: OperationId = UUID.randomUUID()

        val initialState = tracker.state.value

        assertThat(initialState, equalTo(TrackerView.State.empty()))

        val rule1 = stasis.client_android.lib.collection.rules.Rule(
            id = 0,
            operation = stasis.client_android.lib.collection.rules.Rule.Operation.Include,
            directory = "/tmp/1",
            pattern = "*"
        )

        val rule2 = rule1.copy(directory = "/tmp/2")

        val rule3 = rule1.copy(directory = "/tmp/3")

        tracker.backup.specificationProcessed(
            operation = operation,
            unmatched = listOf(
                rule1 to RuntimeException("Test failure"),
                rule2 to RuntimeException("Test failure"),
                rule3 to RuntimeException("Test failure"),
            )
        )

        runBlocking {
            eventually {
                val completedState = tracker.state.await().operations[operation].notNull()

                assertThat(
                    completedState.failures.sorted(),
                    equalTo(
                        listOf(
                            "Rule [+ /tmp/1 *] failed with [Test failure]",
                            "Rule [+ /tmp/2 *] failed with [Test failure]",
                            "Rule [+ /tmp/3 *] failed with [Test failure]",
                        )
                    )
                )
            }
        }
    }

    @Test
    fun trackRecoveryEvents() {
        val tracker = DefaultTracker()

        val operation: OperationId = UUID.randomUUID()
        val file = Paths.get("test").toAbsolutePath()

        val initialState = tracker.state.value

        assertThat(initialState, equalTo(TrackerView.State.empty()))

        tracker.recovery.entityExamined(operation, entity = file, metadataChanged = true, contentChanged = false)
        tracker.recovery.entityExamined(operation, entity = file, metadataChanged = true, contentChanged = true)
        tracker.recovery.entityCollected(operation, entity = file)
        tracker.recovery.entityProcessed(operation, entity = file)
        tracker.recovery.metadataApplied(operation, entity = file)
        tracker.recovery.failureEncountered(operation, failure = RuntimeException("test failure"))
        tracker.recovery.completed(operation)

        runBlocking {
            eventually {
                val completedState = tracker.state.await().operations[operation].notNull()

                assertThat(completedState.completed != null, equalTo(true))

                assertThat(completedState.stages.contains("examination"), equalTo(true))
                assertThat(completedState.stages["examination"]!!.steps.size, equalTo(2))

                assertThat(completedState.stages.contains("collection"), equalTo(true))
                assertThat(completedState.stages["collection"]!!.steps.size, equalTo(1))

                assertThat(completedState.stages.contains("processing"), equalTo(true))
                assertThat(completedState.stages["processing"]!!.steps.size, equalTo(1))

                assertThat(completedState.stages.contains("metadata-applied"), equalTo(true))
                assertThat(completedState.stages["metadata-applied"]!!.steps.size, equalTo(1))

                assertThat(completedState.failures, equalTo(listOf("test failure")))
            }
        }
    }

    @Test
    fun trackServerEvents() {
        val tracker = DefaultTracker()

        val server1 = "test-server-01"
        val server2 = "test-server-02"

        val initialState = tracker.state.value

        assertThat(initialState, equalTo(TrackerView.State.empty()))

        tracker.server.reachable(server1)
        tracker.server.reachable(server2)
        tracker.server.unreachable(server1)

        runBlocking {
            eventually {
                val actual = tracker.state.await().servers.mapValues { it.value.reachable }

                val expected = mapOf(
                    server1 to false,
                    server2 to true
                )

                assertThat(actual, equalTo(expected))
            }
        }
    }

    @Test
    fun provideOperationUpdates() {
        val tracker = DefaultTracker()

        val operation: OperationId = UUID.randomUUID()
        val file = Paths.get("test").toAbsolutePath()

        val initialState = tracker.state.value

        assertThat(initialState, equalTo(TrackerView.State.empty()))

        runBlocking {
            tracker.backup.entityExamined(operation, entity = file, metadataChanged = true, contentChanged = false)

            eventually {
                val operationStateExamined = tracker.operationUpdates(operation).await()
                assertThat(operationStateExamined.stages.size, equalTo(1))
                assertThat(operationStateExamined.completed, equalTo(null))
            }

            tracker.backup.entityCollected(operation = Operation.generateId(), entity = file) // other operation
            tracker.recovery.completed(operation)

            eventually {
                val operationStateCompleted = tracker.operationUpdates(operation).await()
                assertThat(operationStateCompleted.stages.size, equalTo(1))
                assertThat(operationStateCompleted.completed, not(equalTo(null)))
            }
        }
    }

    @get:Rule
    val instantTaskExecutorRule = InstantTaskExecutorRule()
}
